pkg/internal/itest/testpeer/generator.go

```diff
@@ -0,0 +1,138 @@
+package testpeer
+
+import (
+	"context"
+	"time"
+
+	ds "github.com/ipfs/go-datastore"
+	delayed "github.com/ipfs/go-datastore/delayed"
+	ds_sync "github.com/ipfs/go-datastore/sync"
+	blockstore "github.com/ipfs/go-ipfs-blockstore"
+	delay "github.com/ipfs/go-ipfs-delay"
+	bsnet "github.com/ipfs/go-libipfs/bitswap/network"
+	"github.com/ipfs/go-libipfs/bitswap/server"
+	routinghelpers "github.com/libp2p/go-libp2p-routing-helpers"
+	tnet "github.com/libp2p/go-libp2p-testing/net"
+	p2ptestutil "github.com/libp2p/go-libp2p-testing/netutil"
+	"github.com/libp2p/go-libp2p/core/host"
+	peer "github.com/libp2p/go-libp2p/core/peer"
+	mocknet "github.com/libp2p/go-libp2p/p2p/net/mock"
+)
+
+// NewTestPeerGenerator generates a new TestPeerGenerator for the given
+// mocknet
+func NewTestPeerGenerator(ctx context.Context, mn mocknet.Mocknet, netOptions []bsnet.NetOpt, bsOptions []server.Option) TestPeerGenerator {
+	ctx, cancel := context.WithCancel(ctx)
+	return TestPeerGenerator{
+		seq:        0,
+		ctx:        ctx, // TODO take ctx as param to Next, Instances
+		mn:         mn,
+		cancel:     cancel,
+		bsOptions:  bsOptions,
+		netOptions: netOptions,
+	}
+}
+
+// TestPeerGenerator generates new test peers bitswap+dependencies
+// TODO: add graphsync/markets stack, make protocols choosable
+type TestPeerGenerator struct {
+	seq        int
+	mn         mocknet.Mocknet
+	ctx        context.Context
+	cancel     context.CancelFunc
+	bsOptions  []server.Option
+	netOptions []bsnet.NetOpt
+}
+
+// Close closes the clobal context, shutting down all test peers
+func (g *TestPeerGenerator) Close() error {
+	g.cancel()
+	return nil // for Closer interface
+}
+
+// Next generates a new test peer with bitswap + dependencies
+func (g *TestPeerGenerator) Next() TestPeer {
+	g.seq++
+	p, err := p2ptestutil.RandTestBogusIdentity()
+	if err != nil {
+		panic("FIXME") // TODO change signature
```

Comment:
what's going on here, are we tied to a particular API for this struct or is this just an annoying TODO to get to later because it's too cumbersome for now?

Reply:
copied from go-bitswap # oy